### PR TITLE
[Flex Attention] support num_heads > 1 in block_mask

### LINKED ISSUE #148527: [FlexAttention] Error using create_block_mask with mask head number greater than 1
### 🐛 Describe the bug

I have encountered an error when using the flex_attention function in combination with a block mask.  Specifically, the error occurs when the mask `created by create_block_mask` is configured with a head number greater than 1. If the mask head number is set to 1, the code runs without any issues.

To reproduce this problem, I have provided the code snippet below.  This code is self-contained and should directly reproduce the error I am experiencing.

```python
from torch.nn.attention.flex_attention import create_block_mask, flex_attention
import torch

flex_attention = torch.compile(flex_attention, dynamic=False)

q = torch.randn(1, 32, 8, 128, dtype=torch.bfloat16, device="cuda:0")
k = torch.randn(1, 32, 8, 128, dtype=torch.bfloat16, device="cuda:0")
v = torch.randn(1, 32, 8, 128, dtype=torch.bfloat16, device="cuda:0")

def easy_head_attention_mod(head_num):
    head_type = torch.tensor([False if i % head_num == 0 else True for i in range(head_num)], dtype=torch.bool, device=q.device)

    def mask_mod(b, h, q_idx, kv_idx):
        bi_mask = True & head_type[h]
        causal_mask = q_idx >= kv_idx

        return bi_mask & causal_mask

    return mask_mod


mask_mod = easy_head_attention_mod(32) # Error occurs when head_num is greater than 1, e.g., 32
# If head_num is set to 1 (e.g., mask_mod = easy_head_attention_mod(1)), the code runs without error

mask = create_block_mask(mask_mod, 1, 32, 8, 8, device=q.device, _compile=True)

# Use `enable_gqa=True` with corresponding inputs here would bring more bugs
attn_output = flex_attention(q, k, v, block_mask=mask)

print(attn_output.shape)
```

Upon running the code, the following Assertion Error is raised:

```shell
/tmp/torchinductor_root/hp/chpkqukqm77fa3dop4cafoobwxnw5allj4hh2ti37awkdk6gh34c.py:118: unknown: block: [0,3,0], thread: [62,0,0] Assertion `` failed.
/tmp/torchinductor_root/hp/chpkqukqm77fa3dop4cafoobwxnw5allj4hh2ti37awkdk6gh34c.py:118: unknown: block: [0,3,0], thread: [63,0,0] Assertion `` failed.
...
```

Thank you for your time and attention to this issue. I hope this report is helpful in identifying and resolving the problem.

### Versions

torch==2.7.0.dev20250302+cu118

### PR DESCRIPTION
**Pull request by BoyuanFeng**
Previously flex decoding errors when block mask has num_heads > 1. So users have to use num_heads=1, or explicitly mark `kernel_options={"FORCE_USE_FLEX_ATTENTION": True}`.

This PR fixes this issue. When not using grouped query attention (GQA, i.e., Hq == Hkv), we support block mask with num_heads = 1 and num_heads = num_query_heads (i.e., Hq). This is the same setting as flex attention kernel.

When using GQA (i.e., Hq != Hkv), we support block mask with num_heads = 1. When num_heads = Hq, we fall back to flex attention kernel so user don't need to explicitly mark `kernel_options={"FORCE_USE_FLEX_ATTENTION": True}` anymore.

Why fallback? In the current flex decoding triton kernel, grouped query heads for the same kv head are handled by the same thread block. Supporting num_heads = Hq with GQA requires support different kv num blocks for different query heads in the same thread block, leading to lots of redundant workload. So we should better use the main flex_attention kernel where each query head is handled by a separate block.

Fixes #148527
Fixes #147267

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @Chillee @drisspg @yanboliang

### CODE REVIEW

File: torch/_inductor/kernel/flex_attention.py [hunk 1]
```diff
@@ -1325,7 +1341,10 @@ def flex_attention(
         for k, v in kernel_options.items()
     }
     kernel_options.setdefault("FLOAT32_PRECISION", get_float32_precision())
-    if _use_flex_decoding(query, kernel_options):
+    enable_gqa = V.graph.sizevars.evaluate_expr(
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:1344]: don't we pass in "enable_gqa" as an arg to the op?

File: test/inductor/test_flex_decoding.py [hunk 2]
```diff
@@ -961,6 +961,34 @@ def test_non_equal_head_dims(self, dtype, score_mod, head_dims):
             score_mod, dtype, B, Hq, 1, qk_d, B, Hkv, S, V_D=v_d
         )
 
+    @supported_platform
+    @common_utils.parametrize("dtype", test_dtypes_fast)
+    @common_utils.parametrize("score_mod", test_score_mods)
+    @common_utils.parametrize("head_dims", test_Hq_Hkv)
+    def test_head_dependent_mask_mod(self, dtype: torch.dtype, score_mod, head_dims):
```
Comment (drisspg) [inline test/inductor/test_flex_decoding.py:968]: can you also add a test that makes sure we are exercising the dynamic path

**Review by drisspg — APPROVED** [review]
